text/0000-attributes-galore.md

```diff
@@ -0,0 +1,570 @@
+   to work on it; Meanwhile, if `#[cfg(bar)] MyType: Ord` only applied to
+   `MyType` we would get `: Ord` left, which is meaningless and ill-formed.
+
+3. It is consistent with the interpretation of:
+
+   ```rust
+   where
+       for<'a> A: B
+   ```
+
+   which associates as `where for<'a> (A: B)` instead of `where (for<'a> A): B`.
+
+# Prior art
+[prior-art]: #prior-art
+
+## Java
+
+[annotations]: https://en.wikipedia.org/wiki/Java_annotation
+
+Java's [annotations] are a form of syntactic metadata in the same way as
+Rust's `#[attribute]`s are. For example, we may write:
+
+```java
+public class Foo extends Bar {
+    @Override // An annotation; Enforces that `greet` overrides Bar's `greet`.
+    public String greet() { "I am Foo" }
+}
+```
+
+[java8_annotations]: https://docs.oracle.com/javase/tutorial/java/annotations/basics.html
+
+Since Java 8, it is possible to place annotations where a type is expected.
+For example, [we may write][java8_annotations]:
+
+```java
+class UnmodifiableList<T> implements
+    @Readonly List<@Readonly T> { ... }
+
+void monitorTemperature() throws
+    @Critical TemperatureException { ... }
+```
+
+The purpose of these annotations is to provide *"pluggable type systems"*;
+For example, you may write:
+
+```java
+@NonNull String foo = "bar";
+```
+
+## Lifetimes, bounds, and constraints
+
+As far as is known to us, there are not many languages with
+bounds / constraints (Haskell does), or lifetimes (Cyclone ~does)
+which also have annotations or attributes.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+None.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+Having introduced attributes attached to lifetimes, types, bounds,
+and constraints, there would still exist notable places where attributes
+are not yet allowed. Chiefly among these are:
+
++ Expressions:
+
+  ```rust
+  do_stuff(#[foo] 1 + 2);
+  ```
+
++ Patterns:
+
+  ```rust
+  A(#[foo] x) | #[bar] B => ...
+  ```
+
+These are the two places where effort should be directed towards to come up
+with designs that solve any precedence issues that exist.
+
+However, there are also other places where attributes could be allowed;
+For example, we could allow attributes on macro arms:
+
+```rust
+macro_rules! mac {
+    #[foo]
+    ($x:item) => { ... };
+
+    #[bar]
+    ($x:lifetime) => { ... };
+}
+```
+
+This could allow for code generation of macro arms themselves and
+to introduce more macro fragment specifiers, through desugaring semantics,
+without changing the language itself.
+
+Other, more exotic, places where attributes could be allowed are for example:
+
++ in the middle of paths, i.e. `::#[foo] std::#[bar] cell::Cell`
++ in UFCS, i.e. `<Type as #[bar] Trait>::Thing`
++ in visibilities, i.e. `pub(#[foo] crate)`
++ on ABI specifications, i.e. `extern #[foo] "C" ...`
+
+However, the utility of these forms are less clear.
```

Comment:
Here's a use case for `#[cfg]` on macro arms: https://github.com/iliekturtles/uom/pull/62, basically depending on which features are active different types need to have macro arms added for them, and then there can be a fallback arm that is a noop for the types that are not currently active. Because of the lack of this feature instead the macro arms delegate out to another per-type macro that is replaced with a noop if that type is not active.